### ISSUE TITLE
Bug: Oqtane.Maui - Windows - File Manager - Error Message: File Upload Failed Index was outside the bounds of the array.

### ISSUE THREAD

**Issue by thabaum** [oqtane/oqtane.framework#3580]
Running the Oqtane.Maui project in debug mode Visual Studio I am getting an error with the Windows - File Manager 

Error Message: File Upload Failed Index was outside the bounds of the array.

Exception: System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at Oqtane.Modules.Controls.FileManager.UploadFiles()

Properties: {"Error":"Index was outside the bounds of the array."}

Works from Oqtane Server with Server Hosting Model, I have not tested WebAssembly Hosting Model.

I tested this with site settings logo and user profile photo areas same results from file manager in using the Maui Windows app being the above error message.

This was tested using dev branch from a few days ago.

**Comment by sbwalker** [comment]
I am able to reproduce this issue. I am not sure why Blazor Hybrid would have an issue with a standard file upload (ie. input type="file") when it works fine on Blazor Server or Blazor WebAssembly. This will require more investigation.

**Comment by thabaum** [comment]
@sbwalker I have been reviewing the differences in the interop.js files both in the Oqtane.Server and the Oqtane.Maui projects.

Not sure the issue is here but there is differences.  I also tried using <InputFile> element and it seemed to give similar errors.  The other place I have reviewed is the UploadFiles() method in the FileManager where it splits files into arrays and checks files.  Digging deeper would be the API Controller, Service or Repository but I have not seen anything yet that jumps out at me to resolve this even after refactoring things and testing out a few ideas that all ended in similar results, issues with the : splitting I believe in the UploadFiles() method.  I spent some time there but then dug into the interop.js logic which showed differences between the two projects for the same function.

Do we even need to still use javascript or can this be handled in C# and Blazor 100%?

I am trying to create a production IIS environment following the blog but it seems I am having issues with release 5.0.1.  I had prior releases working but maybe something is bugged on my end but it does not seem to load the site app.  I will give this another go this weekend... report back after trying the legacy version 4 out.

**Comment by sbwalker** [comment]
@thabaum I believe you identified the issue in your post above. The .NET MAUI project has its own code to host the share razor components - which also means that the .NET MAUI project must include its own static resources (ie. CSS and JS files). So basically anytime a static file is changed in the Oqtane.Server/wwwroot folder, the same change needs to be applied to the Oqtane.Maui/wwwroot folder. One example is the interop.js file which had some modifications in recent versions to include additional metadata related to the files being uploaded. This is causing the file upload error when running on Oqtane.Maui. There may be other JS/CSS files which need to be reconciled as well.

**Comment by thabaum** [comment]
@sbwalker after further investigation I believe I found the issue which was another idea I had but didnt want to mention until I had more evidence to support.

I believe the issue is relating to CORS policy.  

`Access to XMLHttpRequest at 'http://localhost:44357/api/file/upload' from origin 'https://0.0.0.0' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.`

```
Uncaught ReferenceError: xhr is not defined
    at request.upload.onerror (interop.js:357:75)

error (async) 
 uploadFiles @ interop.js:355

```

```
@ interop.js:361 
  
        POST http://localhost:44357/api/file/upload net::ERR_FAILED 
        uploadFiles @ interop.js:361 
```

I noticed the download upload feature was using the http://localhost:44357 to upload for accessing the api, but the file upload was using https://0.0.0.0

This also causes the logo sometimes not to load as it changes the request for the logo from http to https.

Looks like to resolve this an `Access-Control-Allow-Origin` header may need to be included?  Or issue is with the API Url used to access the file upload api?  Just thinking a bit why this fails so we can get closer to a solution.

I am not sure uploadFiles needs to be included as a javascript file or even if you need one in the Maui project as it will use the one loaded in the browser without overwritten this if you dont include in the index.html file correct?  I will test out removing this file.

**Comment by sbwalker** [comment]
@thabaum I will investigate the CORS issue further however I believe this problem is still related to the problem I reported to Microsoft 2 years ago.  Basically when you are building a Blazor application that is intended to be used in web, desktop, and mobile there is no convenient way to deal with static assets. This is because Blazor Hybrid expects static assets to be local to the device and Blazor Web expects them to be on the server... which makes it cumbersome to handle all use cases using shared components.

https://github.com/dotnet/maui/issues/10898

**Comment by thabaum** [comment]
@sbwalker I seem to have it to where I get 400 errors now (not authorized) but still works via browsers.

Another issue I found is how you can actually drag and a drop file into the browse text area and it almost works, except it brings up a few exceptions along the way to where I need to hit continue in visual studio but it works uploading from browser.  So to bring up an additional error to investigate while working in this area drag and drop.  It will generate the reload error we discussed in another issue which shows more why you need the dismiss button for errors like these.

Drop a file from a folder here:
![image](https://github.com/oqtane/oqtane.framework/assets/13577556/4269685d-5670-4ff0-a9bd-de1cc146e213)

It would be nice to handle the drag and drop of files and have an area to drop them and option to enable/disable it along with capture features if any can be included to enhance the file manager component.  For now at least handle it to where it will not glitch out the browser if a user drag and drops.  I will create a new issue for this if I find a solution specifically for it.  But while looking into CORS issue this can make the list to investigate drag and drop file behavior.

I was wanting to see how we can use a device to capture if there is a camera.  Unfortunately no camera on this system to test with, but I will soon have a setup to test via iPhone, Droid along with Windows

**Comment by sbwalker** [comment]
#3605 updates the interop.js in the .NET MAUI project so that it is consistent with the Oqtane.Server project. This addresses the File Upload Failed Index was outside the bounds of the array error as the uploadFiles() JavaScript method had been modified in recent releases. Please test this fix to determine if it fixes the issue. If you encounter other issues related to CORS, etc... please log a new issue with steps to repro. In regards to drag-and-drop - this was never a supported use case for the FileManager component so should not be considered an "issue".

**Comment by thabaum** [comment]
I had tested out this latest repo and tried those changes as well prior to reporting hoping I would fix it but had no luck.  I have been reviewing ways to rewrite the logic to test things further.

I will check the CORs as well.  I wanted to get this all tested out with production to make sure it is not an IDE issue but I have having all kinds of SSL issues with 500 error if I enable SSL on any sites currently.  I will log this as another issue.

**Comment by sbwalker** [comment]
@thabaum yes, the file upload issue still exists when trying to upload files using the .NET MAUI client. This is because the .NET MAUI Blazor Hybrid app runs on IP 0.0.0.0 and the backend API is running on http://localhost:44357. The .NET HttpClient calls work because they are setting the base address as http://localhost:44357. The XMLHttpRequest used in JavaScript is not setting any base address - which results in the cross origin issue.

**Comment by sbwalker** [comment]
@thabaum I have not had time to test if using an absolute Url in the JavaScript method would resolve this issue. Basically pass a fully qualified Url as the posturl to the uploadFiles() method in interop,js. There are a number of scenarios where absolute Urls need to be used in order for Blazor Hybrid to function properly when sharing razor components.

```
request.open('POST', posturl, true);

```

**Comment by thabaum** [comment]
I am heading in that direction now being sure I log all issues properly.  I will see if I can find any possible solutions or added notes today along the way.

**Comment by thabaum** [comment]
@sbwalker A suggestion to review for Microsoft you may investigate is somehow have a wwwroot folder for the shared project that gets merged into projects with these static resources included?  Maybe it needs to be a different project called SharedStatic or something... idk I dont get paid enough or at all to think about this :) but it would make sense to have something follow that.  There is probably already a request out for it and I am sure they are aware of it and have plenty of ideas.  But it sure would be nice to have in some form.  This would help for things like JS files, however the Client project does not need this... so how to include these using one file that supports this.  Maybe there can be a used shared static resources option on a project and enable it to look for that resource, or not?!

**Comment by sbwalker** [comment]
@thabaum were you able to confirm that the error "File Upload Failed Index was outside the bounds of the array" is resolved and this specific issue can be closed?

**Comment by thabaum** [comment]
@sbwalker I tested this prior to putting up this issue. It does not resolve it.  In fact I been all up and around the JS and File manager logic.  I have tested this with the CORS policy, but not on the methods themselves as discussed.

There is something odd going on, like the fact I cannot drag and drop on the UI like I can.  I think something is blocking the files from uploading?  Maybe even a permissions issue in Windows possibly?  Something I feel is stopping the file from uploading and investigating this has opened up a few other issues.  I have not reported them all yet... as the others are less priority...  pretty much dealing with the warnings after the errors type of list.

**Comment by thabaum** [comment]
@sbwalker I do see the relation between the fileUpload and CORS as I seen the error along with the xrs.status fix I just submitted in the errors.  Hopefully I fixed that properly.

But in discussion relating to CORS and Uploads, I resolved that error CORS, at least to where it was not present and ran into NULL form data.


```
/profile?returnurl=%2F%3Frefresh:1  Access to XMLHttpRequest at 'http://localhost:44357/api/file/upload' from origin 'https://0.0.0.0' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.
localhost:44357/api/file/upload:1 
```

![image](https://github.com/oqtane/oqtane.framework/assets/13577556/caf42df4-247f-4add-8b98-8b113b547c83)


I resolve the CORS issue, but wihtout the CORS issue you dont see the other too which helps really to diagnose a little more here.

1. When CORS issue is resolved, the file is null thus I have another PR to submit to fix this issue because it crashes the app.
  File Controller Method pretty much errors on first IF statement since it does not check for NULL.  I have this issue resolved with checking for null and returning if it is. Currently `formfile` and `folder` are null when attempting to upload.
2. The bigger issue to fix this issue is the second error in the screenshot `Failed to load resource: net::ERR_FAILED`

     Here is another clue:
     
     
![image](https://github.com/oqtane/oqtane.framework/assets/13577556/82583ad8-ae00-4f4e-82d1-0c1e4b5f47da)

access to localhost denied ?  maybe this is just because I clicked on the link in the error but thought I would test access to this "bad request" since it is not really telling me why.

I am getting a few replies going forgive me I am bouncing round so much between issues I may post twice...

**Comment by thabaum** [comment]
Still an issue with file upload in production site setup.

- POST /api/file/upload - 400 Bad Request

I did see a lot of 80 to 443 redirects looping on the server logs, 

- POST /api/Log - 80
- POST /api/Log - 443 

multiple consecutive requests with 307 and 404 
- GET /api/File/name/unchecked.png/4 - 80
- GET /api/File/name/unchecked.png/4 - 443 

Looks like it tried to get the file a few times and quit.

This is from the MAUI app attempt.

I was going to try the browser but after stopping the site to see if it would make the maui app close/crash test the site would not start again with a 500.30 error.

**Comment by sbwalker** [comment]
@thabaum I can no longer get the message "File Upload Failed Index was outside the bounds of the array" to appear. I believe this error message was resolved by synchronizing the interop.js file in the .NET MAUI project with the changes in the Server project. Note that I am not saying that the file upload issue is fixed in .NET MAUI - I am saying that the issue "Error Message: File Upload Failed Index was outside the bounds of the array" ((the title of this issue) is resolved. I am going to close this issue and open a new issue which is focused on the remaining file upload issues.

**Comment by thabaum** [comment]
@sbwalker I was able to get past this error with with CORS if I am not mistaken.  Before I ever put this issue up, I copied the fileUpload both ways to see if one or the other worked while using Maui app.  All my efforts exhausted me so I put up an issue.

Now since then we had that recent issue with saving site settings.  This could of been tampering with my test results.

I believe I uploaded a logo once, then could not upload again.   But I have put CORS support on the framework and upload file controller method to handle this.

I am now at a 400 error antiforgery token error exists. `[Oqtane.Security.AutoValidateAntiforgeryTokenFilter] AutoValidateAntiforgeryTokenFilter Failure For https://cooldomain.com/api/file/upload`

This is a production environment test setup with valid trusted SSL.  Again adding CORS support and then that allowed this next step or error to resolve.  I think I may need to do some updates to the CORS policy I created itself.  It loads only a specific way, you have to include the http:// or https:// or it will not work. 

I will go through my issues tomorrow and clean up the top post with a more clear summary of what is known to help troubleshoot.
